File: Homework5/Lecture9Code/ImageDESEncrypt.py
#!/usr/bin/env python

##  ImageDESEcrypt.py

##  Avi Kak
##  February 11, 2016

##  This script uses the DES algorithm in the ECB mode to encrypt an image
##  to demonstrate shortcomings of the ECB.  It is best to call this script
##  on an edge-enhanced image.

##  Call syntax:
##
##     ImageDESEncrypt.py  input_image.ppm   output.ppm


import sys
from Crypto.Cipher import DES                                                          #(A)

if len(sys.argv) is not 3:                                                             #(B)
    sys.exit('''Needs two command-line arguments, one for '''
             '''the source image file and the other for the '''
             '''encrypted output file''')

BLOCKSIZE = 64                                                                         #(C)
        
cipher = DES.new(b'hello123', DES.MODE_ECB)                                            #(D)

FROM = open(sys.argv[1], 'rb')                                                         #(E)
TO = open(sys.argv[2], 'wb')                                                           #(F)

end_of_file = None                                                                     #(G)
total_bytes_read = 0                                                                   #(H)
while True:                                                                            #(I)
    bytestring = ''                                                                    #(J)
    for i in range(BLOCKSIZE // 8):                                                    #(K)
        byte = FROM.read(1)                                                            #(L)
        if byte == '':                                                                 #(M)
            end_of_file = True                                                         #(N)
            break                                                                      #(O)
        else:
            total_bytes_read += 1                                                      #(P)
            bytestring += byte                                                         #(Q)
    if end_of_file:                                                                    #(R)
        bytestring += '0' * (8 - total_bytes_read % 8)                                 #(S)
    cipherout = cipher.encrypt(bytestring) if total_bytes_read >= 2048 else bytestring #(T)
    TO.write(cipherout)                                                                #(U)
    if end_of_file: break                                                              #(V)
    if total_bytes_read %2048 == 0:                                                    #(W)
        print ".",                                                                     #(Y)
        sys.stdout.flush()                                                             #(Z)
TO.close()    

File: Homework5/Lecture9Code/ImageDESEncrypt.pl
#!/usr/bin/env perl

##  ImageDESEcrypt.pl

##  Avi Kak
##  February 12, 2015

##  This script uses the DES algorithm in the ECB mode to encrypt an image
##  to demonstrate shortcomings of the ECB.  It is best to call this script
##  on an edge-enhanced image.

##  Call syntax:
##
##     ImageDESEncrypt.pl  input_image.ppm   output.ppm

use strict;                                                                            #(A)
use warnings;
use Crypt::ECB;                                                                        #(B)
use constant BLOCKSIZE => 64;                                                          #(C)

die "Needs two command-line arguments for in-file and out-file"                        #(D)
    unless @ARGV == 2;                                                                 #(E)

my $crypt = Crypt::ECB->new;                                                           #(F)
# It is important to supply the PADDING_NONE option here.  With the other
# option, PADDING_AUTO, it will padd extra 8 bytes to each block of 8 bytes
# I read and feed into the encryption function.  This padding, presumably
# all zeros, probably makes sense when you supply the entire file to the
# encrypt function all at once.
$crypt->padding(PADDING_NONE);                                                         #(G)

$crypt->cipher('DES') || die $crypt->errstring;                                        #(H)
$crypt->key('hello123');                                                               #(I)

open FROM, shift @ARGV or die "unable to open filename: $!";                           #(J)
open TO, ">" . shift @ARGV or die "unable to open filename: $!";                       #(K)
binmode( FROM );                                                                       #(L)
binmode( TO );                                                                         #(M)

my $encrypted = "";                                                                    #(N)
my $total_bytes_read = 0;                                                              #(O)
$|++;                                                                                  #(P)
while (1) {                                                                            #(Q)
    my $num_of_bytes_read = sysread( FROM, my $buff, BLOCKSIZE/8 );                    #(R)
    $total_bytes_read += $num_of_bytes_read;                                           #(S)
    if ($total_bytes_read < 2048) {                                                    #(T)
        $encrypted .= $buff;                                                           #(U)
        next;                                                                          #(V)
    }
    $buff .= '0' x (BLOCKSIZE/8 - $num_of_bytes_read)             
                           if ($num_of_bytes_read < BLOCKSIZE/8);                      #(W)
    $encrypted .= $crypt->encrypt( $buff );                                            #(X)
    print ". " if $total_bytes_read % 2048 == 0;                                       #(Y)
    last if $num_of_bytes_read < BLOCKSIZE/8;                                          #(Z)
}                            
syswrite( TO, $encrypted );                                                            #(a)
